plan/plan.go

```diff
@@ -271,14 +271,12 @@ type Rel interface {
 	// result should be 3 columns consisting of the 5th, 2nd and 1st
 	// output columns from the underlying relation.
 	OutputMapping() []int32
-	// Remap utilizes the OutputMapping to construct the expected result
-	// record type from the output RecordType of the underlying Relation.
-	// If the output mapping is nil, then this just returns the input
-	// type.
-	Remap(types.StructType) types.StructType
-	// RecordType returns the output record type of the underlying relation
-	// as a struct type.
-	RecordType() types.StructType
+	// DirectOutputSchema returns the output record type of the underlying
+	// relation as a struct type.
+	DirectOutputSchema() types.StructType
+	// FinalOutputSchema returns the final output schema used by applying
+	// the mapping over the unmapped DirectOutputSchema.
+	FinalOutputSchema() types.StructType
```

Comment:
If we don't export above, we can keep this named RecordType(). Seems like a less disruptive breaking change.

Reply:
RecordType behaves like DirectOutputSchema (often calling Remap to make it behave like FinalOutputSchema).  For almost all uses of the function that's wrong.  I'm in favor of a bigger break.  It will also get folks in the right place to deal with the upcoming Remap change (where we start pushing folks to using the separate Remap/ChangeMapping method).  Without RecordType() there is no reason for folks to call Remap() any longer which frees that name up (eventually) to use instead of ChangeMapping.

Reply:
I'm not clear on where you're disagreeing. Let me be more concrete on how I look at this change:

1. RecordType() should have always returned the final output schema of the relation.
1. It didn't and that was a bug. If you only used emit_kind=direct, the bug had no impact
1. You're fixing the bug.
1. When fixing a bug, we shouldn't need to change the name of the function.
1. DirectOutputSchema() is an internal concern of the relation, and should not be exported.

Where is the disagreement?

Reply:
If you used RecordType() correctly it would be x.Remap(x.RecordType() which would remap the already remapped results.  Nearly all uses of it was as DirectOutputSchema().  In fact all uses of Remap() are with RecordType().

Reply:
As discussed offline, we can keep the name but we should transition to returning a different type (such as RecordType) instead of using types.StructType.  I did try implementing that here but it became way too large of a change.  My next PR will be to update Remap -- hopefully I can do that there.